R/sigfig.R

```diff
@@ -27,82 +27,107 @@ format_decimal <- function(x, sigfig = 3) {
   stopifnot(is.numeric(x))
   sigfig <- check_sigfig(sigfig)
 
-  n <- length(x)
   abs_x <- abs(x)
 
-  # If already bigger than sigfig, can round to zero.
-  # Otherwise ensure we have sigfig digits shown
-  exp <- floor(log10(abs_x))
-  round_x <- signif(abs_x, pmax(sigfig, exp + 1))
-  digits <- ifelse(exp > sigfig, 0, sigfig - exp - ifelse(exp <= 0, 1, 0))
-
-  rhs_digits <- pmax(digits - pmax(exp, 0), 0)
-  dec <- !is.na(digits) & rhs_digits > 0
+  rhs_digits <- compute_rhs_digits(abs_x, sigfig)
+  dec <- rhs_digits > 0
 
   # Do we need negative signs?
   neg <- !is.na(x) & x < 0
-  if (any(neg)) {
-    neg_col <- ifelse(neg, "-", " ")
-  } else {
-    neg_col <- rep("", n)
-  }
-
-  # Digits on LHS of .
-  num <- is.finite(x)
 
+  round_x <- signif(abs_x, pmax(sigfig, compute_exp(abs_x) + 1))
   lhs <- trunc(round_x)
   lhs_zero <- lhs == 0
   lhs_str <- sprintf("%.0f", lhs)
   lhs_width <- max(nchar(lhs_str))
   lhs_sig <- substr(lhs_str, 1, sigfig)
   lhs_non <- substr(lhs_str, sigfig + 1, nchar(lhs_str))
 
+  # Digits on RHS of .
+  rhs <- round_x - lhs
+  rhs_digits <- compute_rhs_digits(abs_x, sigfig)
+  rhs_num <- as.character(abs(round(rhs * 10 ^ rhs_digits)))
+
+  structure(
+    list(
+      neg = format_neg(neg),
```

Comment:
We should probably call these format methods at print time, rather than creation. But that might imply a more significant refactoring than you want to do here.